Add price range filter to Product Collection block

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Adds a Price Range Filter to the Product Collection Block's Inspector Controls
<img src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/62e10296-ec53-49ad-a3ed-a47fbd628b5d" width="300">


- **Integration of Price Range Filter**: A `priceRange` attribute has been added to the block's attribute, allowing merchants to define a price range for displaying products. This is particularly useful for creating more targeted and specific product displays based on pricing criteria.
- **Back-end Support and Filtering Logic**: Corresponding back-end adjustments in `ProductCollection.php` ensure that the product query accounts for the new price range filter, effectively filtering products within the specified price boundaries.
- **Use of 'Auto' for Price Values**: The implementation allows for dynamic price range selection with an 'Auto' placeholder. When the price fields are left blank (auto), the filter dynamically adjusts the range to include all products, offering flexibility in how the price range is applied.

This update significantly improves the versatility of the Product Collection block, empowering merchants to curate products more effectively based on price ranges, and thus enhancing the shopping experience for customers.

Closes #42377

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

#### Case 1: General

1. Insert a product collection block into a new post.
2. From the Filters dropdown, select the “Price Range” filter. It will appear as shown:

<img src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5a9e6a14-4aa1-4c00-abbf-6c3280524f2a" width="300">

3. Adjust the MIN & MAX values. Ensure that the products displayed in the Editor are filtered according to these updated values.
    - **Note:** The changes will only take effect if you *hit ENTER* or *click elsewhere*. This is designed to prevent unnecessary API calls while the merchant is adjusting the value.
4. Save the modifications and confirm that the same updates are visible on the Frontend.
5. Ensure that you can reset the filter's value by clicking in two different places, as indicated in the screenshot below:
<img src="https://github.com/woocommerce/woocommerce-blocks/assets/16707866/d6f2bf7f-0335-4f15-a28b-b618963a5e7a" width="300">

#### Case 2: Prices excluding tax are displayed including tax
**Note:** Before testing, make sure that you have enabled the taxes. Also, set up standard & reduced rate taxes.

- Go to “WooCommerce → Settings → Tax”
    - Set “Prices entered with tax” = “No, I will enter prices exclusive of tax"
    - Set "Display prices in the shop” = “Including tax” as shown in screenshot below

<img src="https://github.com/woocommerce/woocommerce/assets/16707866/b0653454-be8d-49fc-b911-d9c9e6fcfe28" width="400">


- Add Product Collection block to a new post
    - Notice displayed price = Regular/Sale price + Tax
- Now add the Price Range filter to inspector controls
- Change MIN & MAX value. Verify that products are filtered based on displayed price.
    - Also, test decimal values.
- Go to the front end and verify that the price filter works as expected on Frontend.

#### Case 3: Prices including tax are displayed excluding tax

- Go to “WooCommerce → Settings → Tax”
    - Set “Prices entered with tax” = “Yes, I will enter prices inclusive of tax"
    - Set "Display prices in the shop” = “Excluding tax”
- Add Product Collection block to a new post
    - Notice that displayed price = Regular/Sale price - Tax
- Now add the Price Range filter to inspector controls.
- Change MIN & MAX value. Verify that products are filtered based on displayed price.
    - Also, test decimal values.
- Go to frontend, and verify that price filter works as expected on Frontend too.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Product Collection - Add Price Range Filter

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

imanish003 committed:
Add price range filter to Product Collection block

This update introduces a price range filter feature to Product Collection. Changes include:

1. Constants Update:
   - Added `priceRange` as undefined in `DEFAULT_QUERY` and `DEFAULT_FILTERS` in `constants.ts`.

2. Style Adjustments:
   - Added CSS for `.wc-block-product-price-range-control` in `editor.scss` to align the input text to the end.

3. Component Integration:
   - In `inspector-controls/index.tsx`, the `PriceRangeControl` component is now imported and integrated.

4. New Components:
   - `PriceTextField.tsx` and `PriceRangeControl/index.tsx` have been created to handle price range inputs in the Product Collection block.

5. Backend Integration:
   - `PriceRange` interface added in `types.ts` for type support.
   - In `ProductCollection.php`, a filter (`add_price_range_filter`) is added to modify the main query based on the price range, including adjustments for tax considerations.

Overall, this enhancement allows users to filter products within a specific price range. The backend adjustments ensure that the filtering respects tax settings and displays accurate prices.

imanish003 committed:
Fix formatting

imanish003 committed:
Fix: Price range filter not working on Editor

imanish003 committed:
Improve: Share logic between Frontend & Editor

imanish003 committed:
Add changelog

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/42377-product-collection-filters-price-range

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks, woocommerce

imanish003 committed:
Remove duplicate changelog file

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/price-range-control/PriceTextField.tsx:
```diff
@@ -0,0 +1,60 @@
+				<InputControlSuffixWrapper>
+					{ currency?.symbol }
+				</InputControlSuffixWrapper>
+			}
```
There's a `prefix` and `suffix` properties on the `currency` object that takes into account WooCommerce settings where you can choose the currency sign position:

<img width="659" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/8069e390-e08b-4a96-b648-804efc12d07f">

I think it should be taken into account. It seems it's been covered in Shaun's designs with `$` sign as well.

<img width="345" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/4b905ce0-9b63-492e-ad73-9eb81027aa90">

I may be nitpicking here though, so let me know what you think about it.

kmanijak replied:
Similarly, there's `decimalSeparator` that differs between currencies but I see NumberControl doesn't allow to specify this. So I'm mentioning this for the record but I don't think we have to implement it straight away.

imanish003 replied:
Hi Karol,

I've made the following updates:

- Replaced `NumberControl` with `InputControl`, enabling us to incorporate the current symbol with the value. Previously, using `NumberControl` didn't allow invalid numbers as a value pr
- Implemented price formatting logic:
    - Utilizes `currency.minorUnit` for decimal places
    - Applies `currency.thousandSeparator` and `currency.decimalSeparator`
- Tweaked the logic to display the currency symbol as either a prefix or suffix.

With these enhancements, we now have more flexibility, including addressing scenarios where min > max or max < min. What are your thoughts on handling this? Personally, I think it's unnecessary, as merchants can set appropriate min & max values themselves.

Could you please review the changes? The JavaScript code has undergone significant modifications. Thanks 🙏🏻

kmanijak reviewed (commented):
I tested the changes and the filtering is correctly based on the prices displayed in the Product Collection. Thanks for extra effort to address edge cases with different tax settings! 🙌 

I left a comment regarding currency settings. Please let me know what you think of it and I'll either wait for change or approve depending on the outcome.

dinhtungdu commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php:
```diff
@@ -337,23 +344,43 @@ private function get_final_frontend_query( $query, $page = 1, $is_exclude_applie
-		return $this->filter_query_to_only_include_ids( $merged_query, $handpicked_products );
+		$result = $this->filter_query_to_only_include_ids( $merged_query, $handpicked_products );
+
+		if ( isset( $query['priceRange'] ) ) {
+			/**
+			 * We are adding these extra query arguments to be used in `posts_clauses`
+			 * because there are 2 special edge cases we wanna handle for Price range filter:
+			 * Case 1: Prices excluding tax are displayed including tax
+			 * Case 2: Prices including tax are displayed excluding tax
+			 *
+			 * Both of these cases require us to modify SQL query to get the correct results.
+			 */
+			$result = array_merge(
+				$result,
+				array(
+					'isProductCollection' => true,
+					'priceRange'          => $query['priceRange'],
+				)
+			);
+		}
```
I'm a bit concerned about this logic here:
- Because we're now using `priceRange` and `isProductCollection` as query args, should we make them valid query arg in `get_valid_query_vars()`?
- Or, the `isProductCollection` argument is only used with `priceRange`, should we change the `priceRange` arg to something more unique so we can just pass that arg only?

After solving the issue above, I think we can pass the priceRange to the `merge_queries()` method, having to use `array_merge` with `merge_queries` make me a bit confused at the first glance.

imanish003 replied:
Totally see your points! 🙌🏻

> Or, the `isProductCollection` argument is only used with `priceRange`, should we change the `priceRange` arg to something more unique so we can just pass that arg only?
> 

I think it's a good call to stick with `isProductCollection`, as it lets us insert a check [here](https://github.com/woocommerce/woocommerce/blob/97037d847fb7b56a5cc8c85d0b4738b31d7dec7d/plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php#L1079-L1082). We could make the `priceRange` argument name more unique, but clarity wins in this case. This approach also paves the way for more intricate logic down the road. We can examine the `isProductCollection` argument and tweak the WP_Query accordingly. But hey, I'm open to ideas. If you feel renaming `priceRange` is the way to go, let's explore that. 🙂

> having to use `array_merge` with `merge_queries` make me a bit confused at the first glance.
> 

To address this, I have refactored the code in https://github.com/woocommerce/woocommerce/pull/42858/commits/97037d847fb7b56a5cc8c85d0b4738b31d7dec7d. Could you please review the changes and share your feedback? Thanks 🙏🏻

imanish003 committed:
Enhanced Input Control for Price Fields

Key changes include:

1. **Switch to Input Control**: Replaced the NumberControl component with the more versatile InputControl. This offers better handling of currency formatting and user input.

2. **Currency Formatting Logic**: Added robust logic for formatting numbers according to the currency settings. This includes handling thousand separators, decimal places, and currency symbols.

3. **String-to-Number Conversion**: Implemented a function to convert user-entered strings back to numbers, accounting for currency symbols and separators. This ensures accurate parsing of user input for processing.

4. **Input Handling Improvements**: Modified the onChange handlers for minimum and maximum price inputs. Now, they correctly handle edge cases like undefined or zero values, maintaining consistency in the user interface and data processing.

imanish003 committed:
Refactor price range query handling

1. Introduction of a new method `get_price_range_query_args()` to encapsulate the logic for handling price range queries, especially for the two edge cases:
   - Prices excluding tax displayed including tax.
   - Prices including tax displayed excluding tax.

2. Removal of direct conditionals in the `get_query_results()` method, replacing them with a call to the new `get_price_range_query_args()` method. This makes the code more modular and easier to understand.

This refactor enhances readability and maintainability of the code, ensuring that special cases in price range filtering are handled more effectively.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/price-range-control/PriceTextField.tsx:
```diff
@@ -0,0 +1,127 @@
+		// Replace the thousand separator with an empty string and decimal separator with a dot
+		const normalizedValue = valueWithoutCurrencySymbol
+			.replace(
+				new RegExp( `\\${ currency.thousandSeparator }`, 'g' ),
```
While `decimalSeparator` is always there (I think), `thousandSeparator` may not be defined (as an example we don't use a thousand separators in Poland). In such a case this throws an error that crashes block.

```
react-dom.min.js?ver=18.2.0:10 SyntaxError: Invalid regular expression: /\/g: \ at end of pattern (at PriceTextField.tsx:86:5)
    at new RegExp (<anonymous>)
    at convertCurrencyStringToNumber (PriceTextField.tsx:86:5)
    at Object.onChange (PriceTextField.tsx:108:25)
    at Object.onChange [as onChangeHandler] (components.min.js?ver=f6e63a4760dcece8b909:48:202114)
    at components.min.js?ver=f6e63a4760dcece8b909:48:197068
    at Ur (react-dom.min.js?ver=18.2.0:10:73411)
    at Jr (react-dom.min.js?ver=18.2.0:10:80643)
    at Zr (react-dom.min.js?ver=18.2.0:10:80489)
    at Gr (react-dom.min.js?ver=18.2.0:10:80023)
    at react-dom.min.js?ver=18.2.0:10:91526
```

![image](https://github.com/woocommerce/woocommerce/assets/20098064/3e0acde9-c9d8-4bbe-9f37-fce51e4e0924)

Basically if `thousandSeparator` is empty string, RegExp resolves to single slash `\`.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/price-range-control/PriceTextField.tsx:
```diff
@@ -0,0 +1,127 @@
+	value: number;
+	onChange: ( value: number | undefined ) => void;
+	label?: string;
+	suffix?: string;
```
`suffix` seems to be a leftover not used in PriceTextField anymore

```suggestion
```

imanish003 replied:
Good catch! I went ahead and removed it in https://github.com/woocommerce/woocommerce/pull/42858/commits/851923a697f914bb88b731f4786af62c7a15a8e1

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/price-range-control/PriceTextField.tsx:
```diff
@@ -0,0 +1,127 @@
+	val: number | undefined,
+	currency: Currency
+) => {
+	if ( val === undefined || isNaN( val ) ) {
```
Nitpicking, but `isNaN` covers also val being `undefined` so it can be simplified. However I understand if you want to keep it for explicitly showing the cases it's supposed to cover

```suggestion
	if ( isNaN( val ) ) {
```

imanish003 replied:
Actually, I included the `val === undefined` condition because removing it leads to a TypeScript error in the following line:

```ts
let formattedNumber = formatNumber(val, currency);
```

By explicitly adding this condition, I'm informing TypeScript that `val` will not be undefined 😀

kmanijak reviewed (changes requested):
Thanks @imanish003 for addressing a comment regarding currency properties. I left some comments, one major as block may easily crash.

Once I'm really happy you added `decimalSeparator` I'm suggesting we remove the logic with `thousandOperator` completely. Reasons:
1. Thousand operators are easy to read but hard to write and I don't think it's expected to cover that.
2. I checked the logic in Product Editor. In old one thousand operator is even forbidden, in the new one it's not expected but only applied on blur by automatic formatting.
That's why I think it only makes the case more complicated and we can safely remove the logic related to that.

What do you think?

kmanijak commented:
@imanish003 Hmm.. My suggestion about removing thousand operator in https://github.com/woocommerce/woocommerce/pull/42858#pullrequestreview-1790464300 was because of another bug that I spotted.

I thought there's no formatting applied to the input and the operator is used only to properly parse input. But it was because the input didn't format for me on initial blur at all. Only by accident I realised there's expected formatting on blur but it happens only on second blur, please check the video below:

https://github.com/woocommerce/woocommerce/assets/20098064/b3f355eb-e437-428d-97de-fcf1e0d67259

Of course in that case please ignore my suggestion to remove thousand operator, it's actually a great feature and formatting works like in new Product Editor, I really like that but what's required is:
- fix the formatting on blur
- make guards in case there's no `thoousandSeparator` 👍

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/price-range-control/PriceTextField.tsx:
```diff
@@ -0,0 +1,127 @@
+				new RegExp( `\\${ currency.thousandSeparator }`, 'g' ),
+				''
+			)
+			.replace( new RegExp( `\\${ currency.decimalSeparator }` ), '.' );
```
Not sure if that's possible _not to_ define `decimalSeparator` (I assume not 😅 ) but to avoid block crashing in such unlikely scenario I'd add a check if `decimalSeparator` is truthy.

imanish003 replied:
That makes sense. I've implemented additional checks in 79f658d to ensure the block doesn't crash under those circumstances. Thanks for pointing this out 🤝

imanish003 committed:
Remove unnecessary suffix prop

imanish003 committed:
Refactor PriceTextField formatting logic for currency

1. Conditional application of thousand separators: The code now checks for the existence of `currency.thousandSeparator` before applying it. This prevents potential errors when the separator is undefined.

2. Simplified decimal separator handling: Introduced a fallback for the decimal separator, defaulting to a period ('.') if not specified by the currency settings.

3. Enhanced readability and documentation: Added comments to clarify the purpose of code blocks, especially where currency symbols are added or removed, and where value normalization occurs.

4. Function renaming for clarity: Renamed `formatValueWithCurrencySymbol` to `formatCurrency`, which better reflects its purpose.

imanish003 committed:
Fix onBlur issue with PriceTextField component

Refactored the PriceTextField component to utilize useState for better state management. This change introduces a local state variable, 'newValue', to store the current value. The state updates occur in the handleOnChange function, ensuring that the component's state is managed efficiently. Additionally, a new function, handleOnBlur, is implemented to handle the onBlur event, updating the component's state when focus is lost. The handleEnterKeyPress function captures the 'Enter' key press, providing a more user-friendly experience by allowing users to confirm their input with the Enter key.

imanish003 commented:
Hi @kmanijak,

> fix the formatting on blur
> 

I've resolved this issue in the commit https://github.com/woocommerce/woocommerce/pull/42858/commits/dc92a7e756b0ff1ee920e90222063bcdb1f39f4b. I was already aware of it. Originally, I used `isPressEnterToChange` on InputControl, but there seemed to be a bug causing the value not to change on blur. However, it worked correctly when pressing ENTER first, then blurring away.

I initially didn't address this issue, thinking it would be resolved soon in the core. But after your feedback, I agree it's important to fix it now, as users naturally expect it to work on the first blur. So, I:

- Removed `isPressEnterToChange`.
- Implemented custom handling for both blur and Enter key press.

Ideally, I should investigate why it's not working in the core and then contribute upstream. But due to time constraints, I decided to first fix it locally and consider contributing upstream later.

Could you please review it again? Does it look good to you?

kmanijak reviewed (approved):
> I initially didn't address this issue, thinking it would be resolved soon in the core. But after your feedback, I agree it's important to fix it now, as users naturally expect it to work on the first blur.
> Ideally, I should investigate why it's not working in the core and then contribute upstream. But due to time constraints, I decided to first fix it locally and consider contributing upstream later.

I understand where you're coming from. But I agree it's better to at least temporarily fix it locally, good call! 👍 

---

Filter works great now! Thanks for a patience! 🚀

imanish003 committed:
Fix linting error
